Rover: implement steering trim save in armed/manual mode only

Hwurzburg opened the pull request:
uses SAVE_TRIM  rc input channel aux function code and existing Rover steering channel trim save with its range sanity checks (function in radio.cpp)
only enabled while armed and in manual mode

tested in SITL and vehicle...

IamPete1 commented:
We could also implement a plane style auto-trim, plane uses the I term of the PID to offset the trim using a basic low pass filter.

Hwurzburg commented:
the armed qualifier is to be sure the user REALLY wants to change the trim...and you cant judge trim anyway unless moving....
as to trimming the output like plane now does, I considered that but I think the chance of winding up with a very bad manual mode trim after is possible....it works in plane because the only factors affecting pitch/roll trim while in a stabilized mode, are frame trim issues....yaw is not trimmed....but in a boat, while in acro or auto modes has current and wind affecting the yaw trim....so doing it in manual mode gives the user a better chance of correcting frame induced trim only...

rmackay9 commented:
So as discussed on the dev call:

- we need to somehow set the SERVOx_TRIM value for the servo with _FUNCTION = k_steering
- refuse to work for skid-steering vehicles

Hwurzburg commented:
fixed: added !skid_steering qualifier....set servo output trim
so if armed and in manual, non-skid steer, will save steering channels rc value to it trim and set steering output's trim to present pwm value....

SITL tested by inducing trim error at start in steering output (trim not equal to 1500 makes rover turn). then trimed RC in and saved....steering output was back to 1500 and input channel trim changed...will not operate if skid-steering is enabled

Hwurzburg commented:
rebased, added GCS status message

rmackay9 commented on APMrover2/RC_Channel.cpp:
```diff
@@ -223,6 +224,15 @@ void RC_Channel_Rover::do_aux_function(const aux_func_t ch_option, const aux_swi
+    case AUX_FUNC::SAVE_TRIM:
+        if (!rover.g2.motors.have_skid_steering() && rover.arming.is_armed() && rover.control_mode == &rover.mode_manual && ch_flag == HIGH) {
+            SRV_Channels::set_trim_to_servo_out_for(SRV_Channel::k_steering);
+            rover.trim_radio();
```
As discussed on mumble, my feeling at the moment is that we should remove the "rover.trim_radio()" part but I think you're going to give this a test on the water before we decide.

Also maybe we want to change/remove the requirement that the vehicle is in manual mode for this to work.  If we do restrict the mode, and the user is in the wrong mode then we should probably send a text message saying, "Save Trim must be in X mode" or something similar.

Hwurzburg replied:
I tried a version that removes trim.radio and any mode restriction and ran many trials in the wind today. It works the best in ACRO mode with trims averaging ~+/-5us off from best manual trim...then Steering was next best, with RTL/SRTL/GUIDED/AUTO all being up to 25us off randomly,it can be really off if timed during a maneuver or at the end while loitering ...25us is  noticeably incorrect trim when back in MANUAL....of course LOITER and HOLD are no-nos...MANUAL is a bit of a pain having to reset the trim tab, but its obvious that it has to be done...I have pushed a version with out the trim.radio, but disallowing its use during LOITER and HOLD...when merged I will update wiki with suggestion that to get best results use during ACRO ...

rmackay9 commented:
Great, merged!
